[axmszr] iP

axmszr opened the pull request:
# Cbot
### Upgrade your to-do list, today.
---

> `System.out.println("Didn't I do it for you?")` ~Siri

Is speaking too much of a hassle? Does your voice-recognition not recognize your voice? Really wish you could go back to the ol' typing days?
Well, here's the solution you need! Meet **Cbot**, your very own text-based chatbot secretary.
```java
  _____  _            _           |
 /  ___]| |   ^-^   _| |_         |        for (Person p : everyone) {
|  / :D | |___  ___[_ + _]        |            cbot.hug(p);    // cbot says hi!
|  \___ | / . \/ . \ | |          |        }
 \_____]|_,_*_/\_*_/ |_/          |
```
With **Cbot**, you can:
1. Keep track of your to-do list, including deadlines and events!
* *Filter*, *search*, or *sort* your tasks, by:
   - date
   - name
   - type
   - completion status
   - and more!
- [ ] Remain organized and consistent. 😄 🧑‍🎓 📔
- [x] Sleep in peace, knowing **Cbot** will take care of you :")
- [ ] ~Ace your university modules~

___

**Cbot** is easy to learn, and easy to use.

Try it now [here](https://www.youtube.com/watch?v=eN6jkWxxm2Y)[^1]!

[^1]: Sorry, try [this](https://www.youtube.com/watch?v=dQw4w9WgXcQ) instead.

Commit:
Add Gradle support

axmszr committed:
Added Level 1 functionality
Changed name to Cbot (should I change the .java file name too?)

axmszr committed:
Added Level 2 functionality
ToDo class to maintain list
Duke modified

axmszr committed:
Added Level 3 functionality
Task class, as suggested by website
Added some error catching and stuff for commands

axmszr committed:
Renamed Duke.java (main class) to Cbot.java

axmszr committed:
Delete ToDo.class

No idea how it got here oops.

axmszr committed:
Added Level 4 functionality
Made Command an enum, and the corresponding changes to Cbot

axmszr committed:
Added Level 5 functionality
Already had some, just made the no-input case more explicit
Didn't use exceptions

axmszr committed:
Added Level 6 functionality
Not much change! Just added one Command and one in the switch

axmszr committed:
Tried a bit of TextUITesting :P

axmszr committed:
Change tabs to spaces

axmszr committed:
Update gitignore

axmszr committed:
Test force-push tags

axmszr committed:
Remove TEMP TEST.txt

axmszr committed:
Add Level 7 functionality

axmszr committed:
Add Level 8 functionality
Add Filter, Before, After and Sort commands

axmszr committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add Level 7 functionality

# Conflicts:
#	src/main/java/Cbot.java
#	src/main/java/Command.java
#	src/main/java/Deadline.java
#	src/main/java/Event.java
#	src/main/java/TaskList.java

axmszr committed:
Bugfixes from merge
Add done/undone options to Filter

axmszr committed:
Add a lot of OOP (UI, Parser, TimeStuff)
Add custom exceptions

axmszr committed:
Organize into Packages, change too many methods to public
Struggle a bit
Clean up Command

axmszr committed:
Move TaskList to cbot.task package
Clean up import statements
Parser: slightly modify task symbol use

axmszr committed:
Setup IntelliJ
Modify slightly for Java 11 standards

axmszr committed:
Rearrange packages

axmszr committed:
Rearrange packages

axmszr committed:
Add TimeStuff

axmszr committed:
Merge remote-tracking branch 'origin/add-gradle-support'

axmszr committed:
Add Gradle stuff

axmszr committed:
Add JUnit tests
Run JUnit tests with Gradle

axmszr committed:
Add JavaDoc comments to (almost) all non-private classes and methods.

axmszr committed:
Add Level 9 functionality
Add some QoL stuff by allowing not-all-lower case

axmszr committed:
Check Coding Standard - all seems good?

axmszr committed:
Merge branch 'branch-A-JavaDoc' into branch-Level-9

# Conflicts:
#	src/main/java/cbot/io/Command.java
#	src/main/java/cbot/task/Task.java

axmszr committed:
Merge Level-9 and A-JavaDoc
Resolve conflicts

axmszr committed:
Update TaskTest.java to test getDesc()
Build new JAR hehe

axmszr committed:
Use CheckStyle with Gradle

axmszr committed:
Use var-args for Command, Deadline, Event, and Task constructors

axmszr committed:
Add Level 10 functionality
Add GUI classes and package
Add resource images and fxml files

axmszr committed:
Merge branch 'branch-A-CheckStyle' into branch-A-Varargs

# Conflicts:
#	data/cbot_save.txt
#	src/main/java/cbot/Cbot.java
#	src/main/java/cbot/io/Command.java
#	src/main/java/cbot/io/FileStuff.java
#	src/main/java/cbot/io/Parser.java
#	src/main/java/cbot/io/UI.java
#	src/main/java/cbot/task/Deadline.java
#	src/main/java/cbot/task/Event.java
#	src/main/java/cbot/task/TaskList.java

axmszr committed:
Clean up for checkstyle
Set Launcher as the main class
Attempt to merge cleanly (I prefer the CLI one :()

axmszr committed:
Change Task class alphabetical sorting to ignore caps
Try to fix CbotBox.fxml formatting but it's still ugly

Karen-Lam commented on src/main/java/cbot/Cbot.java:
```diff
@@ -0,0 +1,105 @@
+     */
+    public Cbot() {
+        this.fs = new FileStuff(PATH);
+        this.doBye = false;
```
Since doBye holds a boolean variable, I think it might be nice to use a variable name that suggests its a boolean (e.g. isDoBye, isBye)

Karen-Lam commented on src/main/java/cbot/task/TaskList.java:
```diff
@@ -0,0 +1,191 @@
+    //private static final int GAPS = 3;
+    private static final String GAP = "   ";
+
+    private final ArrayList<Task> tdl;
```
I think tdl should be in caps, because it is declared as a final variable

Karen-Lam commented on src/main/java/cbot/task/Task.java:
```diff
@@ -0,0 +1,149 @@
+    public static final String SEP = " ;; ";
+    public static final String TODO_SYMBOL = "T";
+
+    private final String desc;
```
I think desc should be DESC because it is declared as a final variable

Karen-Lam commented on src/main/java/cbot/io/UI.java:
```diff
@@ -0,0 +1,102 @@
+    //CHECKSTYLE.OFF: SingleSpaceSeparator
+
+    // Frequently Used Strings
+    private static final String BLANK     = "   ";
```
I like how the commonly used Strings are all collated at the top of the code. It makes the code look a lot neater.

Karen-Lam commented on src/main/java/cbot/io/Command.java:
```diff
@@ -0,0 +1,163 @@
+     * @return Whether the text means to call this Command.
+     * @see Parser
+     */
+    boolean match(String input) {
```
May be can change the name is isMatch, so it sounds more like a boolean too.

Karen-Lam commented on src/main/java/cbot/io/Command.java:
```diff
@@ -0,0 +1,163 @@
+     * @param text The body text to check.
+     * @return Whether there is missing body text, given the Command type.
+     */
+    boolean missingText(String text) {
```
May be can change the name is isMissingText, so it sounds more like a boolean too.

Karen-Lam reviewed (commented):
Looks great! Found your code was easily to read and follow through. I noted some minor violations in coding standards. but other than that, I think the code is great.

axmszr commented:
Hi @Karen-Lam ! Thank you so much for the comments. I'll definitely look through the booleans and see whether I can name them better!

As for `tdl` and `desc`, I didn't capitalize them because they aren't `static` variables. I think `final` variables don't need to be capitalized by convention? (These names passed the checkstyle. Only `static` variables in camelCase did not.)

axmszr committed:
Change some boolean names to better follow the Java coding standard (as suggested by @Karen-Lam !)

axmszr committed:
Fix issue with text that is too long with TextFlow;
Explicitly split DialogBox into UserBox and CbotBox;
Thank you @hingen for the reference!
Change BLANK to empty String

axmszr committed:
Add // Fallthrough comments in Parser#filter();
TaskList: Change the name of the ArrayList from tdl to tasks.

bokung commented on src/main/java/cbot/io/Parser.java:
```diff
@@ -0,0 +1,362 @@
+                int num = Integer.parseInt(this.text);
+
+                if (tl.notInRange(num)) {
+                    throw new PoorInputException(tl.rangeError(num));
```
This part seems very heavily indented. Perhaps the overall structure can be reorganised?

axmszr replied:
Hi! Thanks so much for the comments.
Could I ask what you're referring to here? Just this `case UNMARK:` or the whole method? I'm not too sure how I can reorganize this `try`-`catch` block :(

bokung commented on src/main/java/cbot/io/Parser.java:
```diff
@@ -0,0 +1,362 @@
+            String msg;
+            ArrayList<String> arrFilter;
+
+            switch (this.text.toLowerCase()) {
```
Is there a point here in including cases with no behaviour if they are gonna fallthrough anyways?

axmszr replied:
Oh! These are just to give the user a bit more flexibility in how they call the filter. For example, both `filter deadline` and `filter dl` (which do the same thing) are accepted and understood by Cbot.
If you know a neater way of doing this, please do let me know!

bokung commented on src/main/java/cbot/io/Parser.java:
```diff
@@ -0,0 +1,362 @@
+                int num = Integer.parseInt(this.text);
+
+                if (tl.notInRange(num)) {
+                    throw new PoorInputException(tl.rangeError(num));
```
Same here with the indentation

bokung commented on src/main/java/cbot/io/Parser.java:
```diff
@@ -0,0 +1,362 @@
+ * Handles much of the parsing of user inputs.
+ */
+public class Parser {
+    private Command c;
```
Perhaps this variable can be renamed to be more descriptive? It is used throughout the entire class and it is not ideal to have to refer back constantly to see what `c` is.

bokung commented on src/main/java/cbot/io/Parser.java:
```diff
@@ -0,0 +1,362 @@
+
+            LocalDateTime bef = TimeStuff.parseDT(this.text);
+
+            ArrayList<String> arrBef = tl.listFilter(t ->
```
This expression seems pretty complicated, perhaps it can be broken down into more intermediate values?

bokung commented on src/main/java/cbot/task/Task.java:
```diff
@@ -0,0 +1,149 @@
+     *
+     * @return Whether the status changed from done to not done.
+     */
+    public boolean noDo() {
```
Here i believe this method should be named to clearly reflect what it's doing. Also, this part can be replaced with the following and it will look much neater.
```java
boolean b = this.isDone;
this.isDone = false;
return b;

```

bokung commented on src/main/java/cbot/task/Task.java:
```diff
@@ -0,0 +1,149 @@
+     *
+     * @return Whether the status changed from not done to done.
+     */
+    public boolean yesDo() {
```
Again this method seems to not reflect it's purpose clearly.

bokung commented on src/main/java/cbot/io/Parser.java:
```diff
@@ -0,0 +1,362 @@
+     * @throws PoorInputException If the input text is improper or erroneous.
+     * @throws DateTimeParseException If some provided datetime is not in a recognized format.
+     */
+    public String respond(TaskList tl) throws PoorInputException, DateTimeParseException {
```
This huge method can perhaps be broken down into smaller functions. It's quite hard to follow the logic of this very long method and you have to indent quite heavily to implement all the logic.

axmszr replied:
Hm, I considered turning each Command into its own Class, and have the functionality be within those, but that felt like too much work hah
Thank you for pointing this out! I'll definitely think about what I can do to make it more readable.

bokung reviewed (commented):
Overall very well documented code! Good use of OOP in structuring your overall program.

axmszr commented:
Hi @bokung ! Thank you so much for the comments. They were all really helpful, and I have ideas on how to improve some of it.

axmszr committed:
Add REMOVE Command (same as DELETE);
Task: Rename yesDo() and noDO() to mark and unmark;
Parser: Rename c to command;
UI: Remove linebreak at the end of printMany.

axmszr committed:
Command: Re-add the linebreaks between enum values.

axmszr committed:
UserBox: Add minHeight="-Infinity";
CbotBox: Brace for reversion of TextFlow to Label.

axmszr committed:
CbotBox: Go back to Label instead of TextFlow/Text, since newline bug fixed.

axmszr committed:
Merge branch 'branch-Label-CbotBox'

# Conflicts:
#	src/main/resources/view/UserBox.fxml

axmszr committed:
Replace UserBox and CbotBox with DialogBox;
Update MainWindow accordingly

axmszr committed:
Change Command#getText() to Command#extractText()

axmszr committed:
Move Parser#respond processing to Command;
Reminder to check for bugs

axmszr committed:
Improve code quality more

axmszr committed:
Move FileStuff and TimeStuff to cbot.util;
Rename UI to Talker (in previous commit)

axmszr committed:
Use Assertions;
Try to assert Command type, but does not work in static context

axmszr committed:
Employ some SLAP for MainWindow#handleUserInput

axmszr committed:
Command: Fix isMissingText error

axmszr committed:
Use Streams for Talker#printMany and some array-dependent TaskList methods

axmszr committed:
Merge pull request #1 from axmszr/branch-A-Assertions

Use Assertions

axmszr committed:
Merge pull request #3 from axmszr/branch-A-CodeQuality

Improve Code Quality

axmszr committed:
Merge branch 'master' into branch-A-Streams

axmszr committed:
Merge pull request #2 from axmszr/branch-A-Streams

Use Streams

axmszr committed:
Fix style of \test\*

axmszr committed:
Set up CI

axmszr committed:
Add C-Update edit functionality;
Add C-MassOps allowance for MARK, UNMARK and DELETE/REMOVE

axmszr committed:
Remove gradle.yml

axmszr committed:
Merge pull request #4 from axmszr/branch-BCD-Extension

Branch BCD Extension

axmszr committed:
Rewrite Command to allow for multiple command names;
Streamline text extraction a bit.

axmszr committed:
Add some QoL fixes to "needs an input" errors

axmszr committed:
Command: improve assert messages

With the recent change to text parsing, it is
more likely for assertions to be violated. This is
especially true for the runCommand functions, where
we want to ensure only the correct Commands are
accessing them.

Let's clarify the assert messages of the runCommand
functions.

If we remove ambiguity from which is which, it will
be easier to debug.

axmszr committed:
Refactor Command

The Command class code is ridiculously long, and a
little messy. It is difficult to find the start of
the static runCommand functions, for example.

Splitting up these sections into their own classes
will neaten things up and improve readability.

Let's,
* extract ThrowingBiFunction from Command into its
own .java file
* extract the static runCommand methods into a
separate class CommandFunction
* re-package all these Command-related classes (as
well as the InputExceptions) into cbot.command

This refactoring will make the code more navigable
and make it easier to add more features in the
future.

axmszr committed:
Follow automated iP code quality feedback;
Task: fix grammar in editDisc JavaDoc;
TaskList: remove dead code;

axmszr committed:
Command: add matches() and extractText() to streamline the abstraction;
CommandFunction: assert correct Command type;
Slightly modify how inputs processed to enable the above.

axmszr committed:
Make more automated JUnit tests.

axmszr committed:
Merge pull request #6 from axmszr/branch-A-MoreTesting

Add final increments

axmszr committed:
Allow the user to see the farewell message for a short while, instead of closing immediately.

axmszr committed:
Merge pull request #7 from axmszr/branch-A-MoreTesting

Allow the user to see the farewell message

axmszr committed:
Add title to application window;
Add UI.png.

axmszr committed:
Oops need to rename the png

axmszr committed:
Add renamed Ui.png

axmszr committed:
Change version to v0.5

axmszr committed:
Add more Command aliases

axmszr committed:
Add a User Guide

axmszr committed:
Cbot: add #hug method
docs/README: change Usage heading to Commands

axmszr committed:
Command: add aft alias for AFTER; reorder BEFORE aliases
docs/README: update accordingly

axmszr committed:
docs/README: mention the different aliases and datetime formats in Features.

axmszr committed:
TimeStuff: fix lower case AM/PM for JDK 11 >:(

axmszr committed:
v0.6

axmszr committed:
Enable assertions in build.gradle
